javalib/src/main/scala/java/util/Arrays.scala

```diff
@@ -12,398 +12,411 @@
 
 package java.util
 
+import java.lang.{reflect => jlr}
+
 import scala.scalajs.js
 
 import scala.annotation.tailrec
 
-import scala.reflect.ClassTag
-
 import ScalaOps._
 
 object Arrays {
 
+  private object NaturalComparator extends Comparator[AnyRef] {
+    @inline
+    def compare(o1: AnyRef, o2: AnyRef): Int =
+      o1.asInstanceOf[Comparable[AnyRef]].compareTo(o2)
+  }
+
+  @inline def ifNullUseNaturalComparator[T <: AnyRef](comparator: Comparator[_ >: T]): Comparator[_ >: T] =
+    if (comparator == null) NaturalComparator
+    else comparator
+
+  /** A custom typeclass for the operations we need in `Arrays` to implement
+   *  the algorithms generically.
+   */
+  private sealed trait ArrayOps[A] {
+    def length(a: Array[A]): Int
+    def get(a: Array[A], i: Int): A
+    def set(a: Array[A], i: Int, v: A): Unit
+  }
+
+  private sealed trait ArrayCreateOps[A] extends ArrayOps[A] {
+    def create(length: Int): Array[A]
+  }
+
+  // ArrayOps instances for reference types
+
+  private sealed trait GenericAnyRefArrayOps[A <: AnyRef] extends ArrayOps[A] {
+    @inline def length(a: Array[A]): Int = a.length
+    @inline def get(a: Array[A], i: Int): A = a(i)
+    @inline def set(a: Array[A], i: Int, v: A): Unit = a(i) = v
+  }
+
   @inline
-  private final implicit def naturalOrdering[T <: AnyRef]: Ordering[T] = {
-    new Ordering[T] {
-      def compare(x: T, y: T): Int = x.asInstanceOf[Comparable[T]].compareTo(y)
-    }
+  private final class SpecificAnyRefArrayOps[A <: AnyRef]
+      extends GenericAnyRefArrayOps[A]
+
+  @inline
+  private implicit def specificAnyRefArrayOps[A <: AnyRef]: ArrayOps[A] =
+    new SpecificAnyRefArrayOps[A]
+
+  @inline
+  private final class ClassArrayOps[A <: AnyRef](clazz: Class[_ <: Array[A]])
+      extends ArrayCreateOps[A] with GenericAnyRefArrayOps[A] {
+    @inline def create(length: Int): Array[A] =
+      createArrayOfClass(clazz, length)
   }
 
-  // Impose the total ordering of java.lang.Float.compare in Arrays
-  private implicit object FloatTotalOrdering extends Ordering[Float] {
-    def compare(x: Float, y: Float): Int = java.lang.Float.compare(x, y)
+  @inline
+  private final class TemplateArrayOps[A <: AnyRef](template: Array[A])
```

Comment:
Instead of this separate class, could we just have a factory for `ClassArrayOps`?

```scala
@inline
private def templateArrayOps[A <: AnyRef](template: Array[A]) =
  new ClassArrayOps(template.getClass())
```

IIUC, this should reduce overall code-size, since a method is much lighter than a class. (and we can inline `createArrayOfClass`).

Reply:
Hrm... I see that if we do this, then we'll not be able to eliminate `getClass` statically, but I'm not sure we ever would need this unless a method asks for `ArrayCreateOps` when `ArrayOps` would be enough.

Reply:
It matters in `sort[T <: AnyRef]`. In that case, we only use the `create` ability when `length > inPlaceSortThreshold`. In the other case, we don't need to create a temporary array, so we avoid the `getClass`.

The code size aspect is not important here because `ClassArrayOps` is always inlined away. So the duplication in `TemplateArrayOps` is not really duplicated. (If we used `ClassArrayOps` instead of `TemplateArrayOps`, then `ClassArrayOps` would be kept where `TemplateArrayOps` currently is, changing nothing.)

Reply:
So, actually this pushed me still a *bit* further. By separating `ArrayCreateOps` from the `ArrayOps` hierarchy, we can achieve to never allocate any array ops instances. The ones that we cannot inline away, we get from static objects, which are allocated once and for all.

I pushed an additional commit with those changes.